File: src/mysqlcookbook/recipes/metadata/affected_rows.pl
#!/usr/bin/perl
# affected_rows.pl

use strict;
use warnings;
use Cookbook;

my $dbh = Cookbook::connect ();

my $stmt = "UPDATE profile SET cats = cats+1 WHERE name = 'Sybil'";

{ # scope begin

# execute statement using do()
#@ _FRAG_1_
my $count = $dbh->do ($stmt);
# report 0 rows if an error occurred
printf "Number of rows affected: %d\n", (defined ($count) ? $count : 0);
#@ _FRAG_1_

} # scope end

{ # scope begin

# execute statement using prepare() plus execute()
#@ _FRAG_2_
my $sth = $dbh->prepare ($stmt);
my $count = $sth->execute ();
printf "Number of rows affected: %d\n", (defined ($count) ? $count : 0);
#@ _FRAG_2_

} # scope end

$dbh->disconnect ();

{ # scope begin

# illustrate use of mysql_client_found_rows
#@ _MYSQL_CLIENT_FOUND_ROWS_
my $conn_attrs = {PrintError => 0, RaiseError => 1, AutoCommit => 1};
my $dsn = "DBI:mysql:cookbook:localhost;mysql_client_found_rows=1";
my $dbh = DBI->connect ($dsn, "cbuser", "cbpass", $conn_attrs);
#@ _MYSQL_CLIENT_FOUND_ROWS_

# this statement changes no rows, but the row count should still
# be nonzero due to the use of mysql_client_found_rows
my $stmt = "UPDATE limbs SET arms = 0 WHERE arms = 0";
my $count = $dbh->do ($stmt);
# report 0 rows if an error occurred
printf "Number of rows affected: %d\n", (defined ($count) ? $count : 0);

} # scope end

File: src/mysqlcookbook/recipes/metadata/affected_rows.php
<?php
# affected_rows.php

require_once "Cookbook.php";

try
{
  $dbh = Cookbook::connect ();
}
catch (PDOException $e)
{
  die ("Error message: " . $e->getMessage () . "\n");
}
try
{
  $stmt = "UPDATE profile SET cats = cats+1 WHERE name = 'Sybil'";
#@ _FRAG_1_
  $count = $dbh->exec ($stmt);
  printf ("Number of rows updated: %d\n", $count);
#@ _FRAG_1_
#@ _FRAG_2_
  $sth = $dbh->prepare ($stmt);
  $sth->execute ();
  printf ("Number of rows updated: %d\n", $sth->rowCount ());
#@ _FRAG_2_
}
catch (PDOException $e)
{
  print ("Error message: " . $e->getMessage () . "\n");
}

$dbh = NULL;

# illustrate how to get rows-matched counts rather than rows-changed counts

try
{
#@ _MYSQL_ATTR_FOUND_ROWS_
  $dsn = "mysql:host=localhost;dbname=cookbook";
  $dbh = new PDO ($dsn, "cbuser", "cbpass",
                  array (PDO::MYSQL_ATTR_FOUND_ROWS => 1));
#@ _MYSQL_ATTR_FOUND_ROWS_
}
catch (PDOException $e)
{
  die ("Error message: " . $e->getMessage () . "\n");
}

try
{
  # this statement changes no rows, but the row count should still
  # be nonzero due to the use of rows-matched counts
  $stmt = "UPDATE limbs SET arms = 0 WHERE arms = 0";
  $count = $dbh->exec ($stmt);
  printf ("Number of rows updated: %d\n", $count);
}
catch (PDOException $e)
{
  print ("Error message: " . $e->getMessage () . "\n");
}
?>
